### PR TITLE
Add prometheus metrics format status with tests

### PR DESCRIPTION
**Pull request by Freggy**
Hello, I followed advice left [here](https://github.com/openresty/lua-resty-upstream-healthcheck/pull/77#issuecomment-1173070778) on #77 and added tests for this functionality.

### CODE REVIEW

File: lib/resty/upstream/healthcheck.lua [hunk 1]
```diff
@@ -651,6 +651,70 @@ local function gen_peers_status_info(peers, bits, idx)
...
+    if not us then
+        return nil, "failed to get upstream names: " .. err
+    end
+
+    local n = #us
+
+    local bits = new_tab(n * 20, 0)
+    local idx = 1
+
+    bits[idx] = "# HELP nginx_upstream_status_info The running staus of nginx upstream"
```
Comment (zhuizhuhaomeng) [inline lib/resty/upstream/healthcheck.lua:685]:
```suggestion
    bits[idx] = "# HELP nginx_upstream_status_info The running status of nginx upstream"
```

File: lib/resty/upstream/healthcheck.lua [hunk 2]
```diff
@@ -651,6 +651,70 @@ local function gen_peers_status_info(peers, bits, idx)
...
+    idx = idx+1
+    bits[idx] = "# TYPE nginx_upstream_status_info gauge"
+
+    for i = 1, n do
+        local u = us[i]
+        local ncheckers = upstream_checker_statuses[u]
+        if not ncheckers or ncheckers == 0 then
+            idx = idx+1
```
Comment (zhuizhuhaomeng) [inline lib/resty/upstream/healthcheck.lua:693]:
```suggestion
            idx = idx + 1
```

File: lib/resty/upstream/healthcheck.lua [hunk 3]
```diff
@@ -651,6 +651,70 @@ local function gen_peers_status_info(peers, bits, idx)
...
+        return nil, "failed to get upstream names: " .. err
+    end
+
+    local n = #us
+
+    local bits = new_tab(n * 20, 0)
+    local idx = 1
+
+    bits[idx] = "# HELP nginx_upstream_status_info The running staus of nginx upstream"
+    idx = idx+1
```
Comment (zhuizhuhaomeng) [inline lib/resty/upstream/healthcheck.lua:686]:
```suggestion
    idx = idx + 1
```

File: lib/resty/upstream/healthcheck.lua [hunk 4]
```diff
@@ -651,6 +651,70 @@ local function gen_peers_status_info(peers, bits, idx)
...
+        local ncheckers = upstream_checker_statuses[u]
+        if not ncheckers or ncheckers == 0 then
+            idx = idx+1
+            bits[idx] = string.format("nginx_upstream_status_info{name=\"%s\",endpoint=\"\",status=\"UNKNOW\",role=\"\"} 1", u)
+            goto continue
+        end
+
+        local peers, err = get_primary_peers(u)
+        if not peers then
+            idx = idx+1
```
Comment (zhuizhuhaomeng) [inline lib/resty/upstream/healthcheck.lua:700]:
```suggestion
            idx = idx + 1
```

**Review by zhuizhuhaomeng — CHANGES_REQUESTED** [review]

File: lib/resty/upstream/healthcheck.lua (hunk 4 above)
Comment (Freggy) [reply lib/resty/upstream/healthcheck.lua]: see https://github.com/openresty/lua-resty-upstream-healthcheck/pull/87/commits/6b0fde522f679a62ce453f77fdfa75dae35523e1

File: lib/resty/upstream/healthcheck.lua (hunk 3 above)
Comment (Freggy) [reply lib/resty/upstream/healthcheck.lua]: see https://github.com/openresty/lua-resty-upstream-healthcheck/pull/87/commits/6b0fde522f679a62ce453f77fdfa75dae35523e1

File: lib/resty/upstream/healthcheck.lua (hunk 2 above)
Comment (Freggy) [reply lib/resty/upstream/healthcheck.lua]: see https://github.com/openresty/lua-resty-upstream-healthcheck/pull/87/commits/6b0fde522f679a62ce453f77fdfa75dae35523e1

File: lib/resty/upstream/healthcheck.lua (hunk 1 above)
Comment (Freggy) [reply lib/resty/upstream/healthcheck.lua]: see https://github.com/openresty/lua-resty-upstream-healthcheck/pull/87/commits/6b0fde522f679a62ce453f77fdfa75dae35523e1

File: lib/resty/upstream/healthcheck.lua [hunk 5]
```diff
@@ -651,6 +651,70 @@ local function gen_peers_status_info(peers, bits, idx)
...
+
+local function gen_peers_prometheus_status_info(peers, bits, idx, u, role)
+    local npeers = #peers
+    for i = 1, npeers do
+        idx = idx + 1
+        local peer = peers[i]
+        if peer.down then
+            bits[idx] = string.format("nginx_upstream_status_info{name=\"%s\",endpoint=\"%s\",status=\"DOWN\",role=\"%s\"} 1", u, peer.name, role)
+        else
+            bits[idx] = string.format("nginx_upstream_status_info{name=\"%s\",endpoint=\"%s\",status=\"UP\",role=\"%s\"} 1", u, peer.name, role)
```
Comment (zhuizhuhaomeng) [inline lib/resty/upstream/healthcheck.lua:663]:
```suggestion
        bits[idx] = string.format("nginx_upstream_status_info{name=\"%s\",endpoint=\"%s\",status=\"%s\",role=\"%s\"} 1", 
                      u, peer.name, peer.down and "DOWN" or "UP" role)
```

**Review by zhuizhuhaomeng — CHANGES_REQUESTED** [review]

File: lib/resty/upstream/healthcheck.lua (hunk 5 above)
Comment (Freggy) [reply lib/resty/upstream/healthcheck.lua]: see https://github.com/openresty/lua-resty-upstream-healthcheck/pull/87/commits/63d01fa4f6cf5b3f2f77f67bdfffff1d1f4bca59

**Comment by Freggy** [comment]
is there something missing still?

**Comment by zhuizhuhaomeng** [comment]
> is there something missing still?

the CI failed.
```
Test Summary Report
-------------------
t/sanity.t (Wstat: 65280 Tests: 105 Failed: 0)
  Non-zero exit status: 255
  Parse errors: Bad plan.  You planned 117 tests but ran 105.
Files=1, Tests=105, 11 wallclock secs ( 0.03 usr  0.00 sys +  0.85 cusr  0.13 csys =  1.01 CPU)
Result: FAIL
The command "prove -r t" exited with 1.
cache.2
store build cache
```

**Comment by zhuizhuhaomeng** [comment]
> is there something missing still?

please move your test to another file 
eg: prometheus.t

**Comment by zhuizhuhaomeng** [comment]
merged with the following patch
```
diff --git a/lib/resty/upstream/healthcheck.lua b/lib/resty/upstream/healthcheck.lua
index 5fcacd2..0099a59 100644
--- a/lib/resty/upstream/healthcheck.lua
+++ b/lib/resty/upstream/healthcheck.lua
@@ -657,7 +657,7 @@ local function gen_peers_prometheus_status_info(peers, bits, idx, u, role)
     for i = 1, npeers do
         idx = idx + 1
         local peer = peers[i]
-        bits[idx] = string.format("nginx_upstream_status_info{name=\"%s\",endpoint=\"%s\",status=\"%s\",role=\
"%s\"} 1", u, peer.name, peer.down and "DOWN" or "UP", role)   
+        bits[idx] = string.format("nginx_upstream_status_info{name=\"%s\",endpoint=\"%s\",status=\"%s\",role=\
"%s\"} 1", u, peer.name, peer.down and "DOWN" or "UP", role)
     end
     return idx
 end
diff --git a/t/prometheus.t b/t/prometheus.t
new file mode 100644
index 0000000..32fdb0b
--- /dev/null
+++ b/t/prometheus.t
@@ -0,0 +1,221 @@
+# vim:set ft= ts=4 sw=4 et fdm=marker:
+use lib 'lib';
+use Test::Nginx::Socket::Lua;
+use Cwd qw(cwd);
+
...
```

**Pull request closed without merging**